pyproject-fmt/rust/src/tests/dependency_groups_tests.rs

```diff
@@ -0,0 +1,141 @@
+use common::taplo::formatter::{format_syntax, Options};
+use common::taplo::parser::parse;
+use common::taplo::syntax::SyntaxElement;
+use indoc::indoc;
+use rstest::rstest;
+
+use crate::dependency_groups::fix;
+use common::table::Tables;
+
+fn evaluate(start: &str, keep_full_version: bool) -> String {
+    let root_ast = parse(start).into_syntax().clone_for_update();
+    let count = root_ast.children_with_tokens().count();
+    let mut tables = Tables::from_ast(&root_ast);
+    fix(&mut tables, keep_full_version);
+    let entries = tables
+        .table_set
+        .iter()
+        .flat_map(|e| e.borrow().clone())
+        .collect::<Vec<SyntaxElement>>();
+    root_ast.splice_children(0..count, entries);
+    let opt = Options {
+        column_width: 1,
+        ..Options::default()
+    };
+    format_syntax(root_ast, opt)
+}
+
+#[rstest]
+#[case::no_groups(
+        indoc ! {r""},
+        "\n",
+        false,
+)]
+#[case::single_group_single_dep(
+        indoc ! {r#"
+    [dependency-groups]
+    test=["a>1.0.0"]
+    "#},
+        indoc ! {r#"
+    [dependency-groups]
+    test = [
+      "a>1",
+    ]
+    "#},
+        false,
+)]
+#[case::single_group_single_dep_full_version(
+        indoc ! {r#"
+    [dependency-groups]
+    test=["a>1.0.0"]
+    "#},
+        indoc ! {r#"
+    [dependency-groups]
+    test = [
+      "a>1.0.0",
+    ]
+    "#},
+        true,
+)]
+#[case::single_group_multiple_deps(
+        indoc ! {r#"
+    [dependency-groups]
+    test=["b==2.0.*", "a>1"]
+    "#},
+        indoc ! {r#"
+    [dependency-groups]
+    test = [
+      "a>1",
+      "b==2.0.*",
+    ]
+    "#},
+        false,
+)]
+#[case::multiple_groups(
+        indoc ! {r#"
+    [dependency-groups]
+    example=["c<1"]
+    docs=["b==1"]
+    test=["a>1"]
+    dev=["d>=2"]
+    "#},
+        indoc ! {r#"
+    [dependency-groups]
+    dev = [
+      "d>=2",
+    ]
+    test = [
+      "a>1",
+    ]
+    docs = [
+      "b==1",
+    ]
+    example = [
+      "c<1",
+    ]
+    "#},
+        false,
+)]
+#[case::include_single_group(
+        indoc ! {r#"
+    [dependency-groups]
+    docs=["b==1"]
+    test=["a>1",{include-group="docs"}]
+    "#},
+        indoc ! {r#"
+    [dependency-groups]
+    test = [
+      "a>1",
+      { include-group = "docs" },
```

Comment:
There is the ability to include other groups in the PEP: https://peps.python.org/pep-0735/#dependency-group-include

These aren't reordered right now, and doing it seems above my Rust skillset... I think it would be nice to have though. Any preference on how to place them? I was thinking:

- include groups should be at the end
- ordered by group name

Thoughts?

Reply:
agreed

Reply:
Actually, upon further inspection, if an dependency group contains an `include-group`, the ordering of the array doesn't happen at all. I tracked this down to this section:

https://github.com/tox-dev/toml-fmt/blob/764b4656da29aecd34c0dc266c9075d825888976/common/src/array.rs#L102-L105

That happens in the "common" array sorting logic, and I'm a bit relunctant on adding something specific to a) pyproject.toml and b) the `dependency-groups` table.

What's the intention with https://github.com/tox-dev/toml-fmt/pull/12? I see that the README mentions `ruff-toml-fmt` and `tox-toml-fmt` but I don't see these projects (yet)?

Reply:
Yeah, that is a work in progress.